src/server.ts

```diff
@@ -16,6 +19,86 @@ import mime from 'mime-types';
 import fs from 'fs/promises';
 import { z } from 'zod';
 
+const execPromise = util.promisify(exec);
+
+export const serverRunId = randomUUID();
+export const activeSandboxContainers = new Map<string, number>();
+
+const containerTimeoutSeconds = parseInt(
+  process.env.NODE_CONTAINER_TIMEOUT || '3600',
+  10
+);
+const containerTimeoutMilliseconds = isNaN(containerTimeoutSeconds)
+  ? 3600 * 1000
+  : containerTimeoutSeconds * 1000;
+
+export async function forceStopContainer(containerId: string): Promise<void> {
+  console.log(`Attempting to stop and remove container: ${containerId}`);
+  try {
+    await execPromise(`docker stop ${containerId} || true`);
+    await execPromise(`docker rm -f ${containerId} || true`);
+    console.log(`Successfully stopped and removed container: ${containerId}`);
+  } catch (error) {
+    console.error(
+      `Error during forced stop/remove of container ${containerId}:`,
+      typeof error === 'object' &&
+        error !== null &&
+        ('stderr' in error || 'message' in error)
+        ? (error as { stderr?: string; message?: string }).stderr ||
+            (error as { message: string }).message
+        : String(error)
+    );
+  } finally {
+    activeSandboxContainers.delete(containerId);
+  }
+}
+
+const scavengerInterval = setInterval(() => {
+  const now = Date.now();
+  console.log(
+    `[Scavenger] Checking ${activeSandboxContainers.size} active containers for timeout (${containerTimeoutSeconds}s)...`
+  );
+  for (const [
+    containerId,
+    creationTimestamp,
+  ] of activeSandboxContainers.entries()) {
+    if (now - creationTimestamp > containerTimeoutMilliseconds) {
+      console.warn(
+        `[Scavenger] Container ${containerId} timed out. Forcing removal.`
+      );
+      forceStopContainer(containerId);
+    }
+  }
+}, 60 * 1000);
+
+async function gracefulShutdown(signal: string) {
+  console.log(`
+Received ${signal}. Starting graceful shutdown...`);
+
+  clearInterval(scavengerInterval);
+
+  const containersToClean = Array.from(activeSandboxContainers.keys());
```

Comment:
We can expose (from the same `containerUtils` file) a function called `cleanActiveContainers` and call it here. It might be useful if we want to add more features to the shutdown in the future